internal/js_library/js_library.bzl

```diff
@@ -46,23 +48,66 @@ def write_amd_names_shim(actions, amd_names_shim, targets):
                 amd_names_shim_content += "define(\"%s\", function() { return %s });\n" % n
     actions.write(amd_names_shim, amd_names_shim_content)
 
-def _js_library(ctx):
-    return [
+def _impl(ctx):
+    if not ctx.files.srcs:
+        fail("No srcs specified")
+
+    source_files = ctx.files.srcs[0].is_source
+    for src in ctx.files.srcs:
+        if src.is_source != source_files:
+            fail("Mixing of source and generated files not allowed")
+
+    sources_depset = depset(ctx.files.srcs)
+
+    result = [
         DefaultInfo(
-            files = depset(ctx.files.srcs),
+            files = sources_depset,
             runfiles = ctx.runfiles(files = ctx.files.srcs),
         ),
         AmdNamesInfo(names = ctx.attr.amd_names),
     ]
 
-js_library = rule(
-    implementation = _js_library,
+    if ctx.attr.package_name:
+        if source_files:
+            path = "/".join([p for p in [ctx.label.workspace_root, ctx.label.package] if p])
+        else:
+            path = "/".join([p for p in [ctx.bin_dir.path, ctx.label.workspace_root, ctx.label.package] if p])
+        result.append(LinkablePackageInfo(
+            package_name = ctx.attr.package_name,
+            path = path,
+            files = sources_depset,
+        ))
+
+    return result
+
+_js_library = rule(
+    implementation = _impl,
     attrs = {
-        "srcs": attr.label_list(allow_files = [".js"]),
+        "package_name": attr.string(),
+        "srcs": attr.label_list(allow_files = True),
         "amd_names": attr.string_dict(doc = _AMD_NAMES_DOC),
-        "module_from_src": attr.bool(),
-        # Used to determine module mappings
+        # module_name for legagy ts_library module_mapping support
+        # TODO: remove once legacy module_mapping is removed
         "module_name": attr.string(),
-        "module_root": attr.string(),
     },
 )
+
+def js_library(
```

Comment:
TODO to remove the macro after breaking change rolls?

Reply:
👍